src/services/service.ts

```diff
@@ -287,6 +288,74 @@ export default class Service {
         return parsed_results
     }
 
+    protected serviceStream<T>(options: ReportOptions, pre_report_hook: PreReportHook) {
+        const query = this.buildCustomerReportQuery(options)
+
+        const call = this.streamSearchData(query)
+
+        return this.streamGenerator<T>(call)
+    }
+
+    private async *streamGenerator<T>(call: ClientReadableStream<SearchGoogleAdsStreamResponse>) {
+        let done = false
+        const accumulator: T[] = []
+
+        function createNewPromise() {
+            let resolveMe = () => {
+                console.log('this should never happen')
+            }
+
+            let rejectMe = (err: Error) => {
+                console.log('this should never happen', err)
+            }
+
+            const p = new Promise((resolve, reject) => {
+                resolveMe = resolve
+                rejectMe = reject
+            })
+
+            return { p, resolveMe, rejectMe }
+        }
+
+        let next_chunk_arrived = createNewPromise()
+
+        call.on('data', (chunk: SearchGoogleAdsStreamResponse.AsObject) => {
+            const results = this.parseServiceResults(chunk.resultsList)
+            for (const item of results) {
+                accumulator.push(item as T)
+            }
+            next_chunk_arrived.resolveMe()
+            next_chunk_arrived = createNewPromise()
+        })
+
+        call.on('end', () => {
+            done = true
+            next_chunk_arrived.resolveMe()
+        })
+
+        call.on('error', (err: Error) => {
+            next_chunk_arrived.rejectMe(err)
+        })
+
+        try {
+            while (!done || accumulator.length) {
+                if (accumulator.length !== 0) {
+                    const p = accumulator.shift()
+                    if (p === undefined) {
+                        throw new Error('UNDEFINED_STREAM_ERROR')
```

Comment:
if it throws is the while loop terminated?

Reply:
yes